File: typescript/src/traceparent.ts
/**
 * W3C Trace Context (traceparent) propagation.
 * Format: 00-{trace_id_32hex}-{parent_id_16hex}-{flags_2hex}
 *
 * This enables distributed tracing across service boundaries.
 * When a traced function makes HTTP calls, the traceparent header
 * is injected so downstream services can correlate their spans.
 */

import { AsyncLocalStorage } from "async_hooks";

// Per-async-context trace context. Concurrent traces each get their own store, so one
// trace's context can't leak into another's outbound requests. The module-level globals
// below remain as a fallback for the legacy imperative setTraceContext/clearTraceContext API.
const traceContextStore = new AsyncLocalStorage<{ traceId: string; spanId: string }>();

let _currentTraceId: string | null = null;
let _currentSpanId: string | null = null;

/**
 * Run `fn` with an isolated trace context. Outgoing-request helpers (getTraceparent /
 * injectTraceparent) called within `fn` (including across awaits) see this context, with
 * no cross-contamination between concurrent traces.
 */
export function withTraceContext<T>(traceId: string, spanId: string, fn: () => T): T {
  return traceContextStore.run(
    { traceId: traceId.replace(/-/g, ""), spanId: spanId.replace(/-/g, "").slice(0, 16) },
    fn,
  );
}

/** Set the active trace context for outgoing requests (legacy imperative API). */
export function setTraceContext(traceId: string, spanId: string): void {
  // Convert UUID format to 32-hex (remove dashes)
  _currentTraceId = traceId.replace(/-/g, "");
  // Take first 16 chars of span ID as parent span
  _currentSpanId = spanId.replace(/-/g, "").slice(0, 16);
}

/**
 * Imperative ALS context for the recorder's start()/end(), which have no callback scope to wrap.
 * Uses AsyncLocalStorage.enterWith so the context is isolated PER ASYNC EXECUTION — two concurrent
 * imperative traces (each in its own async context) get their own store instead of clobbering a
 * shared module global. This is the fix for the old setTraceContext()-writes-a-global contamination:
 * the recorder no longer touches _currentTraceId at all; the globals remain ONLY for the explicitly
 * legacy, single-context setTraceContext() public API.
 */
export function enterTraceContext(traceId: string, spanId: string): void {
  traceContextStore.enterWith({ traceId: traceId.replace(/-/g, ""), spanId: spanId.replace(/-/g, "").slice(0, 16) });
}

/** Clear the imperative ALS context for the current execution (empty store → getTraceparent returns
 *  null WITHOUT falling through to the legacy globals). */
export function exitTraceContext(): void {
  traceContextStore.enterWith({ traceId: "", spanId: "" });
}

/** Clear the active trace context. */
export function clearTraceContext(): void {
  _currentTraceId = null;
  _currentSpanId = null;
}

/** Get the current traceparent header value, or null if no active trace. */
export function getTraceparent(): string | null {
  const scoped = traceContextStore.getStore();
  const traceId = scoped?.traceId ?? _currentTraceId;
  const spanId = scoped?.spanId ?? _currentSpanId;
  if (!traceId || !spanId) return null;
  // version-trace_id-parent_id-flags (01 = sampled)
  return `00-${traceId}-${spanId}-01`;
}

/**
 * Inject traceparent into a headers object (for fetch/axios/http calls).
 * Returns the headers with traceparent added if a trace is active.
 */
export function injectTraceparent(headers: Record<string, string> = {}): Record<string, string> {
  const tp = getTraceparent();
  if (tp) {
    headers["traceparent"] = tp;
  }
  return headers;
}

/**
 * Parse an incoming traceparent header.
 * Returns { traceId, parentId, sampled } or null if invalid.
 */
export function parseTraceparent(header: string): { traceId: string; parentId: string; sampled: boolean } | null {
  const parts = header.split("-");
  if (parts.length !== 4 || parts[0] !== "00") return null;
  if (parts[1].length !== 32 || parts[2].length !== 16) return null;
  return {
    traceId: parts[1],
    parentId: parts[2],
    sampled: parts[3] === "01",
  };
}

File: python/src/retrace/traceparent.py
"""W3C Trace Context (traceparent) propagation for distributed tracing.

Format: 00-{trace_id_32hex}-{parent_id_16hex}-{flags_2hex}

When a traced function makes HTTP calls, inject the traceparent header
so downstream services can correlate their spans with the parent trace.
"""
from typing import Dict, Optional, Tuple
import contextvars

# Per-context trace context (mirrors the _dispatch ContextVar). Each thread / async task resolves
# its own value, so concurrent traces can't leak context into one another's outbound requests.
_trace_ctx: contextvars.ContextVar[Optional[Tuple[str, str]]] = contextvars.ContextVar(
    "retrace_trace_ctx", default=None
)


def set_trace_context(trace_id: str, span_id: str):
    """Set the active trace context for outgoing requests. Returns a token for reset()."""
    return _trace_ctx.set((trace_id.replace("-", ""), span_id.replace("-", "")[:16]))


def clear_trace_context(token=None) -> None:
    """Clear the active trace context. Pass the token from set_trace_context to restore the
    enclosing context (supports nesting); otherwise the context is cleared outright."""
    if token is not None:
        try:
            _trace_ctx.reset(token)
            return
        except (ValueError, LookupError):
            pass
    _trace_ctx.set(None)


def get_traceparent() -> Optional[str]:
    """Get the current traceparent header value, or None if no active trace."""
    ctx = _trace_ctx.get()
    if not ctx:
        return None
    trace_id, span_id = ctx
    if not trace_id or not span_id:
        return None
    return f"00-{trace_id}-{span_id}-01"


def inject_traceparent(headers: Optional[Dict[str, str]] = None) -> Dict[str, str]:
    """Inject traceparent into a headers dict for outgoing HTTP requests."""
    if headers is None:
        headers = {}
    tp = get_traceparent()
    if tp:
        headers["traceparent"] = tp
    return headers


def parse_traceparent(header: str) -> Optional[Tuple[str, str, bool]]:
    """Parse an incoming traceparent header.

    Returns (trace_id, parent_id, sampled) or None if invalid.
    """
    parts = header.split("-")
    if len(parts) != 4 or parts[0] != "00":
        return None
    if len(parts[1]) != 32 or len(parts[2]) != 16:
        return None
    return (parts[1], parts[2], parts[3] == "01")
